File: srcs/server/agents/builtin/read.go
package builtin

import (
	"context"
	"encoding/json"
	"io"
	"os"
)

// FileReadTool definition
var FileReadTool = Tool{
	Name:        "Read",
	Description: "Read a file from the local filesystem.",
	SearchHint: "built-in tool",
	RequiresApproval: false,
	Parameters: json.RawMessage(`{
		"type": "object",
		"properties": {
			"file_path": {
				"type": "string",
				"description": "The absolute path to the file to read."
			}
		},
		"required": ["file_path"]
	}`),
	Execute: func(ctx context.Context, args json.RawMessage) (string, error) {
		var input struct {
			FilePath string `json:"file_path"`
		}
		if err := json.Unmarshal(args, &input); err != nil {
			return "", err
		}

		file, err := os.Open(input.FilePath)
		if err != nil {
			return "", err
		}
		defer file.Close()

		content, err := io.ReadAll(file)
		if err != nil {
			return "", err
		}
		return string(content), nil
	},
}

File: srcs/server/agents/builtin/tools/read.cc
#include "srcs/server/agents/builtin/tools/tool.h"

#include <cerrno>
#include <cstring>
#include <fstream>
#include <sstream>
#include <string>

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "nlohmann/json.hpp"

namespace ohc::agent {

Tool MakeReadTool() {
  return Tool{
      .name        = "Read",
      .description = "Read a file from the local filesystem.",
      .parameters  = nlohmann::json::parse(R"({
        "type": "object",
        "properties": {
          "file_path": {
            "type": "string",
            "description": "The absolute path to the file to read."
          }
        },
        "required": ["file_path"]
      })"),
      .execute = [](const nlohmann::json& args) -> absl::StatusOr<std::string> {
        const auto it = args.find("file_path");
        if (it == args.end() || !it->is_string()) {
          return absl::InvalidArgumentError("Read: missing 'file_path'");
        }
        const std::string path = it->get<std::string>();

        std::ifstream file(path, std::ios::binary);
        if (!file.is_open()) {
          return absl::NotFoundError(
              absl::StrCat("Read: cannot open '", path, "': ",
                           std::strerror(errno)));
        }
        std::ostringstream ss;
        ss << file.rdbuf();
        return ss.str();
      },
  };
}

}  // namespace ohc::agent
